File: chapter02/python/nextGreaterElement.py
def next_greater_element(nums):
    res = nums.copy()
    stack = []
    for i in range(len(res) - 1, -1, -1):
        while len(stack) > 0 and stack[-1] <= nums[i]:
            stack.pop()
        res[i] = -1 if len(stack) == 0 else stack[-1]
        stack.append(nums[i])
    return res


# 假设数组是个环形，可以将idx加倍, 用的时候取idx = idx % len(nums)
def next_greater_element_circle(nums):
    n = len(nums)
    res = nums.copy()
    stack = []
    for i in range(2 * n - 1, -1, -1):
        while len(stack) > 0 and stack[-1] <= nums[i % n]:
            stack.pop()
        res[i % n] = -1 if len(stack) == 0 else stack[-1]
        stack.append(nums[i % n])
    return res


def next_greater_element_idx(nums):
    res = list(range(len(nums)))
    stack = []
    for i in range(len(res) - 1, -1, -1):
        while len(stack) > 0 and nums[stack[-1]] <= nums[i]:
            stack.pop()
        res[i] = -1 if len(stack) == 0 else stack[-1]
        stack.append(i)
    return res


if __name__ == "__main__":
    nums = [2, 1, 2, 4, 3]
    print(next_greater_element_circle(nums))

File: chapter02/cpp/nextGreaterElement.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <stdlib.h>
#include <stack>

using namespace std;

vector<int> next_greater_element(const vector<int>& nums)
{   
    int n = nums.size();
    vector<int> res(n, 0);
    stack<int> s;
    for(int i=n-1; i>=0; i--)
    {   
        while(s.size() > 0 && s.top() < nums[i])
            s.pop();
            
        res[i] = s.size() > 0 ? s.top() : -1;
        s.push(nums[i]);
    }
    return res;
}

vector<int> next_greater_element_circle(const vector<int>& nums)
{   
    int n = nums.size();
    vector<int> res(n, 0);
    stack<int> s;
    for(int i = 2 * n - 1; i >= 0; i--)
    {   
        while(s.size() > 0 && s.top() <= nums[i%n])
            s.pop();
            
        res[i%n] = s.size() > 0 ? s.top() : -1;
        s.push(nums[i%n]);
    }
    return res;
}

vector<int> next_greater_element_idx(const vector<int>& nums)
{   
    int n = nums.size();
    vector<int> res(n, 0);
    stack<int> s;
    for(int i = n - 1; i >= 0; i--)
    {   
        while(s.size() > 0 && nums[s.top()] <= nums[i])
            s.pop();
            
        res[i] = s.size() > 0 ? s.top() : -1;
        s.push(i);
    }
    return res;
}

int main()
{
    vector<int> nums = {2, 1, 2, 4, 3};
    vector<int> res = next_greater_element_idx(nums);
    for(auto& x: res)
        cout << x << endl;
    return 0;
}